x-pack/plugin/ml-feature-index-builder/src/main/java/org/elasticsearch/xpack/ml/featureindexbuilder/job/FeatureIndexBuilderIndexer.java

```diff
@@ -7,188 +7,119 @@
+            Map<String, Object> initialPosition) {
+        super(executor, initialState, initialPosition, new FeatureIndexBuilderJobStats());
 
         this.job = job;
-        this.client = client;
-        logger.info("delete pivot-reviews");
-        
     }
 
-    public synchronized void start() {
-        deleteIndex(client);
-
-        createIndex(client);
-        
-        int runs = 0;
-
-        Map<String, Object> after = null;
-        logger.info("start feature indexing");
-        SearchResponse response;
-        
-        try {
-            response = runQuery(client, after);
-
-            CompositeAggregation compositeAggregation = response.getAggregations().get("feature");
-            after = compositeAggregation.afterKey();
-
-            while (after != null) {
-                indexBuckets(compositeAggregation);
-
-                ++runs;
-                response = runQuery(client, after);
-
-                compositeAggregation = response.getAggregations().get("feature");
-                after = compositeAggregation.afterKey();
-                
-                //after = null;
-            }
-            
-            indexBuckets(compositeAggregation);
-        } catch (InterruptedException | ExecutionException e) {
-            logger.error("Failed to build feature index", e);
-        }
-
-        logger.info("Finished feature indexing");
+    @Override
+    protected String getJobId() {
+        return job.getConfig().getId();
     }
 
-    private void indexBuckets(CompositeAggregation compositeAggregation) {
-        BulkRequest bulkIndexRequest = new BulkRequest();
-        try {
-            for (Bucket b : compositeAggregation.getBuckets()) {
+    @Override
+    protected void onStartJob(long now) {
+    }
 
-                InternalAvg avgAgg = b.getAggregations().get("avg_rating");
+    @Override
+    protected IterationResult<Map<String, Object>> doProcess(SearchResponse searchResponse) {
+        final CompositeAggregation agg = searchResponse.getAggregations().get("feature");
+        return new IterationResult<>(processBuckets(agg), agg.afterKey(), agg.getBuckets().isEmpty());
+    }
 
-                XContentBuilder builder;
+    private List<IndexRequest> processBuckets(CompositeAggregation agg) {
+        return agg.getBuckets().stream().map(b -> {
+            InternalAvg avgAgg = b.getAggregations().get("avg_rating");
+            XContentBuilder builder;
+            try {
                 builder = jsonBuilder();
+
                 builder.startObject();
                 builder.field("reviewerId", b.getKey().get("reviewerId"));
                 builder.field("avg_rating", avgAgg.getValue());
                 builder.endObject();
-                bulkIndexRequest.add(new IndexRequest(PIVOT_INDEX, DOC_TYPE).source(builder));
-
+            } catch (IOException e) {
+                throw new RuntimeException(e);
```

Comment:
There is a class called `UncheckedIOException` that might be better here, as it's clearer to the handling code that it should be unwrapped.